File: backend/app/schemas/system.py
from apiflask import Schema
from apiflask.fields import Integer, String, List, Boolean, Nested, DateTime, Dict
from apiflask.validators import Length, Email

# --- Permission Schemas ---

class PermissionOutSchema(Schema):
    id = Integer()
    name = String(metadata={'description': '权限标识', 'example': 'product:create'})
    description = String(metadata={'description': '权限描述'})

# --- Permission Tree Schemas ---

class PermissionActionSchema(Schema):
    id = Integer(required=True)
    code = String(required=True) # e.g. product:create
    label = String(required=True) # e.g. 创建

class PermissionGroupSchema(Schema):
    name = String(required=True)   # e.g. 商品列表
    label = String(required=True)  # e.g. 商品列表
    permissions = List(Nested(PermissionActionSchema))

class PermissionModuleSchema(Schema):
    name = String(required=True)   # e.g. 商品中心
    label = String(required=True)  # e.g. 商品中心
    children = List(Nested(PermissionGroupSchema))

# --- Data Permission Schemas ---

class DataPermMetaSchema(Schema):
    id = Integer()
    key = String()
    label = String()
    type = String()
    description = String()
    children = List(Nested(lambda: DataPermMetaSchema())) # Recursive

class RoleDataPermConfigSchema(Schema):
    category_key = String(required=True)
    target_user_ids = List(Integer(), load_default=[])
    # Map of resource_key -> scope_type ('all', 'custom')
    resource_scopes = Dict(keys=String(), values=String(), load_default={})

class RoleDataPermBulkSchema(Schema):
    # List of configs
    configs = List(Nested(RoleDataPermConfigSchema), required=True)

# --- Field Permission Schemas ---

class FieldPermMetaSchema(Schema):
    id = Integer()
    module = String()
    field_key = String()
    label = String()
    description = String()

class RoleFieldPermConfigSchema(Schema):
    field_key = String(required=True)
    is_visible = Boolean(load_default=True)
    condition = String(load_default='none') # 'none', 'follower'

class RoleFieldPermUpdateSchema(Schema):
    configs = List(Nested(RoleFieldPermConfigSchema), required=True)

# --- Role Schemas ---

class RoleBaseSchema(Schema):
    name = String(required=True, validate=Length(min=2, max=50), metadata={'description': '角色名称', 'example': 'editor'})
    description = String(metadata={'description': '角色描述'})

class RoleCreateSchema(RoleBaseSchema):
    permission_ids = List(Integer(), load_default=[], metadata={'description': '关联的权限ID列表', 'example': [1, 2]})

class RoleUpdateSchema(RoleBaseSchema):
    name = String(validate=Length(min=2, max=50)) # Optional in update usually, but here we keep validation
    permission_ids = List(Integer(), load_default=[])

class RoleOutSchema(RoleBaseSchema):
    id = Integer()
    permissions = List(Nested(PermissionOutSchema), metadata={'description': '拥有的权限列表'})

class RoleUserAddSchema(Schema):
    user_ids = List(Integer(), required=True, metadata={'description': '用户ID列表', 'example': [1, 2]})

# --- User Schemas ---

class UserBaseSchema(Schema):
    username = String(required=True, validate=Length(min=3, max=50), metadata={'description': '用户名'})
    nickname = String(load_default='', metadata={'description': '昵称/中文名', 'example': '张三'})
    email = String(required=True, validate=Email(), metadata={'description': '邮箱'})
    is_active = Boolean(load_default=True, metadata={'description': '是否激活'})

class UserCreateSchema(UserBaseSchema):
    password = String(required=True, validate=Length(min=6), metadata={'description': '密码'})
    role_ids = List(Integer(), load_default=[], metadata={'description': '关联的角色ID列表'})

class UserUpdateSchema(Schema):
    username = String(validate=Length(min=3, max=50))
    nickname = String(metadata={'description': '昵称'})
    email = String(validate=Email())
    password = String(validate=Length(min=6), load_default=None)
    is_active = Boolean()
    role_ids = List(Integer())

class UserOutSchema(UserBaseSchema):
    id = Integer()
    nickname = String(metadata={'description': '昵称'})
    created_at = DateTime()
    roles = List(Nested(RoleOutSchema(only=['id', 'name', 'description'])), metadata={'description': '所属角色'})

File: frontend/apps/web-antd/src/api/core/system.ts
import { requestClient } from '#/api/request';

/**
 * Permission Definitions
 */
export interface PermissionItem {
  id: number;
  name: string;
  description: string;
}

export interface PermissionAction {
  id: number;
  code: string;
  label: string;
}

export interface PermissionGroup {
  name: string;
  label: string;
  permissions: PermissionAction[];
}

export interface PermissionModule {
  name: string;
  label: string;
  children: PermissionGroup[];
}

/**
 * Data Permission Definitions
 */
export interface DataPermMeta {
  id: number;
  key: string;
  label: string;
  type: 'category' | 'module' | 'resource';
  description?: string;
  children: DataPermMeta[];
}

export interface RoleDataPermConfig {
  category_key: string;
  target_user_ids: number[];
  resource_scopes: Record<string, 'all' | 'custom'>;
}

/**
 * Field Permission Definitions
 */
export interface FieldPermMeta {
  id: number;
  module: string;
  field_key: string;
  label: string;
  description?: string;
}

export interface RoleFieldPermConfig {
  field_key: string;
  is_visible: boolean;
  condition: string; // 'none', 'follower'
}

/**
 * Role Definitions
 */
export interface RoleItem {
  id: number;
  name: string;
  description: string;
  permissions: PermissionItem[];
}

export interface RoleCreateParams {
  name: string;
  description?: string;
  permission_ids: number[];
}

export interface RoleUpdateParams extends RoleCreateParams {}

/**
 * User Definitions
 */
export interface UserItem {
  id: number;
  username: string;
  realname?: string;
  nickname?: string;
  mobile?: string;
  email: string;
  is_active: boolean;
  created_at: string;
  roles: RoleItem[];
}

export interface UserCreateParams {
  username: string;
  email: string;
  password?: string;
  realname?: string;
  nickname?: string;
  mobile?: string;
  role_ids: number[];
  is_active?: boolean;
}

export interface UserUpdateParams extends Partial<UserCreateParams> {}

export interface UserListParams {
  page: number;
  per_page: number;
}

export interface UserListResult {
  items: UserItem[];
  total: number;
  page: number;
  per_page: number;
  pages: number;
}

/**
 * API Methods
 */

// Permissions
export const getPermissionsApi = () => {
  return requestClient.get<PermissionItem[]>('/v1/system/permissions');
};

export const getPermissionTreeApi = () => {
  return requestClient.get<PermissionModule[]>('/v1/system/permissions/tree');
};

// Data Permissions
export const getDataPermissionMetasApi = () => {
  return requestClient.get<DataPermMeta[]>('/v1/system/data-permission-metas');
};

export const getRoleDataPermissionApi = (roleId: number, categoryKey: string) => {
  return requestClient.get<RoleDataPermConfig>(`/v1/system/roles/${roleId}/data-permissions`, {
    params: { category_key: categoryKey }
  });
};

export const saveRoleDataPermissionApi = (roleId: number, data: RoleDataPermConfig) => {
  return requestClient.post<RoleDataPermConfig>(`/v1/system/roles/${roleId}/data-permissions`, data);
};

export const saveRoleDataPermissionsBulkApi = (roleId: number, configs: RoleDataPermConfig[]) => {
  return requestClient.post<void>(`/v1/system/roles/${roleId}/data-permissions/bulk`, { configs });
};

// Field Permissions
export const getFieldPermissionMetasApi = () => {
  return requestClient.get<FieldPermMeta[]>('/v1/system/field-permission-metas');
};

export const getRoleFieldPermissionsApi = (roleId: number) => {
  return requestClient.get<RoleFieldPermConfig[]>(`/v1/system/roles/${roleId}/field-permissions`);
};

export const saveRoleFieldPermissionsApi = (roleId: number, configs: RoleFieldPermConfig[]) => {
  return requestClient.post<RoleFieldPermConfig[]>(`/v1/system/roles/${roleId}/field-permissions`, { configs });
};

// Roles
export const getRolesApi = () => {
  return requestClient.get<RoleItem[]>('/v1/system/roles');
};

export const createRoleApi = (data: RoleCreateParams) => {
  return requestClient.post<RoleItem>('/v1/system/roles', data);
};

export const updateRoleApi = (id: number, data: RoleUpdateParams) => {
  return requestClient.put<RoleItem>(`/v1/system/roles/${id}`, data);
};

export const deleteRoleApi = (id: number) => {
  return requestClient.delete<void>(`/v1/system/roles/${id}`);
};

// Users
export const getUsersApi = (params: UserListParams) => {
  return requestClient.get<UserListResult>('/v1/system/users', { params });
};

export const createUserApi = (data: UserCreateParams) => {
  return requestClient.post<UserItem>('/v1/system/users', data);
};

export const updateUserApi = (id: number, data: UserUpdateParams) => {
  return requestClient.put<UserItem>(`/v1/system/users/${id}`, data);
};

export const deleteUserApi = (id: number) => {
  return requestClient.delete<void>(`/v1/system/users/${id}`);
};

// Role Users
export const getRoleUsersApi = (roleId: number, params?: any) => {
  return requestClient.get<UserListResult>(`/v1/system/roles/${roleId}/users`, { params });
};

export const addRoleUsersApi = (roleId: number, userIds: number[]) => {
  return requestClient.post(`/v1/system/roles/${roleId}/users`, { user_ids: userIds });
};

export const removeRoleUserApi = (roleId: number, userId: number) => {
  return requestClient.delete(`/v1/system/roles/${roleId}/users/${userId}`);
};
